File: app/src/main/java/knf/kuma/directory/DirectoryService.java
package knf.kuma.directory;

import android.annotation.SuppressLint;
import android.app.IntentService;
import android.app.Notification;
import android.app.NotificationManager;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.media.AudioAttributes;
import android.os.Handler;
import android.os.Looper;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;
import android.support.v4.content.ContextCompat;
import android.util.Log;

import org.jsoup.HttpStatusException;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;

import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

import knf.kuma.R;
import knf.kuma.commons.BypassUtil;
import knf.kuma.commons.Network;
import knf.kuma.commons.SSLSkipper;
import knf.kuma.database.CacheDB;
import knf.kuma.database.dao.AnimeDAO;
import knf.kuma.jobscheduler.DirUpdateJob;
import knf.kuma.pojos.AnimeObject;
import knf.kuma.pojos.DirectoryPage;
import pl.droidsonroids.jspoon.Jspoon;

public class DirectoryService extends IntentService {
    public static final int STATE_PARTIAL = 0;
    public static final int STATE_FULL = 1;
    public static final int STATE_FINISHED = 2;
    public static final int STATE_INTERRUPTED = 3;
    public static int NOT_CODE = 5598;
    public static String CHANNEL = "directory_update";
    private static boolean RUNNING = false;
    private long CURRENT_TIME = System.currentTimeMillis();
    private static MutableLiveData<Integer> liveStatus = new MutableLiveData<>();
    private NotificationManager manager;
    private int count = 0;
    private int page = 0;

    public DirectoryService() {
        super("Directory update");
    }

    public static boolean isRunning() {
        return RUNNING;
    }

    public static void run(Context context) {
        if (!RUNNING)
            ContextCompat.startForegroundService(context, new Intent(context, DirectoryService.class));
    }

    public static boolean isDirectoryFinished(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context).getBoolean("directory_finished", false);
    }

    public static LiveData<Integer> getLiveStatus() {
        return liveStatus;
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        startForeground(NOT_CODE, getStartNotification());
        RUNNING = true;
        if (!Network.isConnected()) {
            stopSelf();
            cancelForeground();
        }
        manager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        AnimeDAO animeDAO = CacheDB.INSTANCE.animeDAO();
        SSLSkipper.skip();
        Jspoon jspoon = Jspoon.create();
        setStatus(STATE_PARTIAL);
        doPartialSearch(jspoon, animeDAO);
        setStatus(STATE_FULL);
        doFullSearch(jspoon, animeDAO);
        cancelForeground();
    }

    @SuppressLint("ApplySharedPref")
    private void doPartialSearch(Jspoon jspoon, AnimeDAO animeDAO) {
        final Set<String> strings = PreferenceManager.getDefaultSharedPreferences(this).getStringSet("failed_pages", new LinkedHashSet<>());
        final Set<String> newStrings = new LinkedHashSet<>();
        int partialCount = 0;
        if (strings.size() == 0)
            Log.e("Directory Getter", "No pending pages");
        for (final String s: new LinkedHashSet<>(strings)) {
            partialCount++;
            if (!Network.isConnected()) {
                Log.e("Directory Getter", "Processed " + partialCount + " pages before disconnection");
                stopSelf();
                return;
            }
            try {
                Document document = Jsoup.connect("https://animeflv.net/browse?order=added&page=" + s).cookies(BypassUtil.getMapCookie(this)).userAgent(BypassUtil.userAgent).get();
                if (document.select("article").size() != 0) {
                    List<AnimeObject> animeObjects = jspoon.adapter(DirectoryPage.class).fromHtml(document.outerHtml()).getAnimes(this, animeDAO, jspoon, new DirectoryPage.UpdateInterface() {
                        @Override
                        public void onAdd() {
                            count++;
                            updateNotification();
                        }

                        @Override
                        public void onError() {
                            Log.e("Directory Getter", "At page: " + s);
                            if (!newStrings.contains(s))
                                newStrings.add(String.valueOf(s));
                        }
                    });
                    if (animeObjects.size() > 0)
                        animeDAO.insertAll(animeObjects);
                }
            } catch (Exception e) {
                Log.e("Directory Getter", "Page error: " + s);
                if (!newStrings.contains(String.valueOf(s)))
                    newStrings.add(String.valueOf(s));
            }
        }
        PreferenceManager.getDefaultSharedPreferences(this).edit().putStringSet("failed_pages", newStrings).commit();
    }

    private void doFullSearch(Jspoon jspoon, AnimeDAO animeDAO) {
        page = 0;
        boolean finished = false;
        final Set<String> strings = PreferenceManager.getDefaultSharedPreferences(this).getStringSet("failed_pages", new LinkedHashSet<String>());
        while (!finished) {
            if (!Network.isConnected()) {
                Log.e("Directory Getter", "Processed " + page + " pages before disconnection");
                stopSelf();
                return;
            }
            try {
                Document document = Jsoup.connect("https://animeflv.net/browse?order=added&page=" + page).cookies(BypassUtil.getMapCookie(this)).userAgent(BypassUtil.userAgent).get();
                if (document.select("article").size() != 0) {
                    page++;
                    List<AnimeObject> animeObjects = jspoon.adapter(DirectoryPage.class).fromHtml(document.outerHtml()).getAnimes(this, animeDAO, jspoon, new DirectoryPage.UpdateInterface() {
                        @Override
                        public void onAdd() {
                            count++;
                            updateNotification();
                        }

                        @Override
                        public void onError() {
                            Log.e("Directory Getter", "At page: " + page);
                            if (!strings.contains(String.valueOf(page)))
                                strings.add(String.valueOf(page));
                        }
                    });
                    if (animeObjects.size() > 0) {
                        animeDAO.insertAll(animeObjects);
                    } else if (PreferenceManager.getDefaultSharedPreferences(this).getBoolean("directory_finished", false)) {
                        Log.e("Directory Getter", "Stop searching at page " + page);
                        cancelForeground();
                        break;
                    }
                } else {
                    finished = true;
                    Log.e("Directory Getter", "Processed " + page + " pages");
                    PreferenceManager.getDefaultSharedPreferences(this).edit().putBoolean("directory_finished", true).apply();
                    PreferenceManager.getDefaultSharedPreferences(this).edit().putStringSet("failed_pages", strings).apply();
                    DirUpdateJob.schedule(this);
                    setStatus(STATE_FINISHED);
                }
            } catch (HttpStatusException e) {
                finished = true;
                setStatus(STATE_INTERRUPTED);
            } catch (Exception e) {
                Log.e("Directory Getter", "Page error: " + page);
                if (!strings.contains(String.valueOf(page)))
                    strings.add(String.valueOf(page));
            }
        }
        cancelForeground();
    }

    private void cancelForeground() {
        RUNNING = false;
        stopForeground(true);
        notCancel(NOT_CODE);
    }

    private void updateNotification() {
        Notification notification = new NotificationCompat.Builder(this, CHANNEL)
                .setOngoing(true)
                .setContentTitle("Creando directorio")
                .setContentText("Agregados: " + count)
                .setPriority(NotificationCompat.PRIORITY_MIN)
                .setSmallIcon(R.drawable.ic_directory_not)
                .setColor(Color.parseColor("#e53935"))
                .setWhen(CURRENT_TIME)
                .setSound(null)
                .build();
        notShow(NOT_CODE, notification);
    }

    private Notification getStartNotification() {
        return new NotificationCompat.Builder(this, "directory_update")
                .setOngoing(true)
                .setContentTitle("Verificando directorio")
                .setPriority(NotificationCompat.PRIORITY_MIN)
                .setSmallIcon(R.drawable.ic_directory_not)
                .setSound(null, AudioAttributes.CONTENT_TYPE_SONIFICATION)
                .setColor(Color.parseColor("#e53935"))
                .setWhen(CURRENT_TIME)
                .build();
    }

    private void setStatus(int status) {
        new Handler(Looper.getMainLooper()).post(() -> liveStatus.setValue(status));
    }

    private void notShow(int code, Notification notification) {
        if (manager != null)
            manager.notify(code, notification);
    }

    private void notCancel(int code) {
        if (manager != null)
            manager.cancel(code);
    }

    @Override
    public void onTaskRemoved(Intent rootIntent) {
        RUNNING = false;
        stopForeground(true);
        notCancel(NOT_CODE);
        super.onTaskRemoved(rootIntent);
    }

    public interface OnDirStatus {
        void onFinished();
    }
}

File: app/src/main/java/knf/kuma/directory/DirectoryService.kt
package knf.kuma.directory

import android.annotation.SuppressLint
import android.app.IntentService
import android.app.Notification
import android.app.NotificationManager
import android.content.Context
import android.content.Intent
import android.media.AudioManager
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.core.content.ContextCompat
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.preference.PreferenceManager
import knf.kuma.BuildConfig
import knf.kuma.R
import knf.kuma.commons.*
import knf.kuma.database.CacheDB
import knf.kuma.database.dao.AnimeDAO
import knf.kuma.download.FileAccessHelper
import knf.kuma.download.foreground
import knf.kuma.download.service
import knf.kuma.jobscheduler.DirUpdateWork
import knf.kuma.pojos.AnimeObject
import knf.kuma.pojos.DirectoryPage
import knf.kuma.widgets.emision.WEmisionProvider
import org.jsoup.HttpStatusException
import pl.droidsonroids.jspoon.Jspoon
import java.util.*
import kotlin.properties.Delegates

class DirectoryService : IntentService("Directory update") {
    private val CURRENT_TIME = System.currentTimeMillis()
    private var manager: NotificationManager? = null
    private var count = 0
    private var page = 0
    private var maxAnimes = 3200
    private val TAG = "Directory Getter"
    private var needCookies by Delegates.notNull<Boolean>()

    private val keyFailedPages = "failed_pages"

    private val startNotification: Notification
        get() {
            val notification = NotificationCompat.Builder(this, "directory_update")
                    .setOngoing(true)
                    .setPriority(NotificationCompat.PRIORITY_MIN)
                    .setSmallIcon(R.drawable.ic_directory_not)
                    .setSound(null, AudioManager.STREAM_NOTIFICATION)
                    .setColor(ContextCompat.getColor(this, EAHelper.getThemeColor()))
                    .setWhen(CURRENT_TIME)
            if (PrefsUtil.collapseDirectoryNotification)
                notification.setSubText("Verificando directorio")
            else
                notification.setContentTitle("Verificando directorio")
            return notification.build()
        }

    override fun onCreate() {
        foreground(NOT_CODE, startNotification)
        super.onCreate()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        foreground(NOT_CODE, startNotification)
        return super.onStartCommand(intent, flags, startId)
    }

    override fun onHandleIntent(intent: Intent?) {
        foreground(NOT_CODE, startNotification)
        if (!Network.isConnected) {
            cancelForeground()
            stopSelf()
            return
        }
        needCookies = BypassUtil.isCloudflareActive(BypassUtil.testLink)
        isRunning = true
        setStatus(STATE_VERIFYING)
        manager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val animeDAO = CacheDB.INSTANCE.animeDAO()
        count = animeDAO.count
        SSLSkipper.skip()
        val jspoon = Jspoon.create()
        calculateMax()
        setStatus(STATE_PARTIAL)
        doPartialSearch(jspoon, animeDAO)
        setStatus(STATE_FULL)
        doEcchiRemove(animeDAO)
        checkDownloaded(animeDAO)
        doFullSearch(jspoon, animeDAO)
        doEmissionRefresh(jspoon, animeDAO)
        cancelForeground()
    }

    private fun calculateMax() {
        noCrash {
            val main = jsoupCookiesDir("https://animeflv.net/browse",needCookies).get()
            val lastPage = main.select("ul.pagination li:matches(\\d+)").last().text().trim().toInt()
            val last = jsoupCookiesDir("https://animeflv.net/browse?page=$lastPage",needCookies).get()
            maxAnimes = ((24 * (lastPage - 1)) + last.select("article").size) - 1
            Log.e(TAG, "Max pages = $maxAnimes")
        }
    }

    private fun doEcchiRemove(animeDAO: AnimeDAO) {
        if (BuildConfig.BUILD_TYPE == "playstore" || PrefsUtil.isFamilyFriendly) {
            animeDAO.nukeEcchi()
        }
    }

    private fun checkDownloaded(animeDAO: AnimeDAO) {
        val jspoon = Jspoon.create()
        FileAccessHelper.downloadExplorerCreator.createLinksList().forEach {
            if (!animeDAO.existLink("%${it.substringAfterLast(".net")}")){
                try {
                    val response = okHttpCookies(it).execute(followRedirects = true)
                    val body = response.body?.string()
                    if (response.code == 200 && body != null) {
                        val webInfo = jspoon.adapter(AnimeObject.WebInfo::class.java).fromHtml(body)
                        animeDAO.insert(AnimeObject(it, webInfo))
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
                if (needCookies)
                    Thread.sleep(5000)
            }
        }
    }

    private fun doEmissionRefresh(jspoon: Jspoon, animeDAO: AnimeDAO) {
        try {
            animeDAO.allLinksInEmission.forEach {
                try {
                    val animeObject = AnimeObject(it, jspoon.adapter(AnimeObject.WebInfo::class.java).fromHtml(jsoupCookiesDir(it,needCookies).get().outerHtml()))
                    val current = animeDAO.getAnimeByAid(animeObject.aid)
                    if (current == null || current != animeObject)
                        animeDAO.updateAnime(animeObject)
                    WEmisionProvider.update(this)
                } catch (e: Exception) {
                    return@forEach
                }
            }
        }catch (e:Exception){
            //
        }
    }

    @SuppressLint("ApplySharedPref")
    private fun doPartialSearch(jspoon: Jspoon, animeDAO: AnimeDAO) {
        val strings = PreferenceManager.getDefaultSharedPreferences(this).getStringSet(keyFailedPages, LinkedHashSet())
        val newStrings = LinkedHashSet<String>()
        var partialCount = 0
        if (strings?.size == 0)
            Log.e(TAG, "No pending pages")
        for (s in LinkedHashSet(strings)) {
            partialCount++
            if (!Network.isConnected) {
                Log.e(TAG, "Processed $partialCount pages before disconnection")
                stopSelf()
                return
            }
            try {
                if (needCookies)
                    Thread.sleep(6000)
                else
                    Thread.sleep(1000)
                val document = jsoupCookiesDir("https://animeflv.net/browse?order=added&page=$s", needCookies).get()
                if (document.select("article").size != 0) {
                    val animeObjects = jspoon.adapter(DirectoryPage::class.java).fromHtml(document.outerHtml()).getAnimes(animeDAO, jspoon, object : DirectoryPage.UpdateInterface {
                        override fun onAdd() {
                            count++
                            updateNotification()
                        }

                        override fun onError() {
                            Log.e(TAG, "Error at page: $s")
                            if (!newStrings.contains(s))
                                newStrings.add(s.toString())
                        }
                    }, needCookies)
                    if (animeObjects.isNotEmpty())
                        animeDAO.insertAll(animeObjects)
                }
            } catch (e: HttpStatusException) {
                if (e.statusCode == 403 || e.statusCode == 503) {
                    setStatus(STATE_INTERRUPTED)
                    stopSelf()
                    cancelForeground()
                }
            } catch (e: Exception) {
                Log.e(TAG, "Page error: $s | ${e.message}")
                if (!newStrings.contains(s.toString()))
                    newStrings.add(s.toString())
            }

        }
        PreferenceManager.getDefaultSharedPreferences(this).edit().putStringSet(keyFailedPages, newStrings).commit()
    }

    private fun doFullSearch(jspoon: Jspoon, animeDAO: AnimeDAO) {
        page = 1
        var finished = false
        val strings = PreferenceManager.getDefaultSharedPreferences(this).getStringSet(keyFailedPages, LinkedHashSet())
        while (!finished) {
            if (!Network.isConnected) {
                Log.e(TAG, "Processed $page pages before disconnection")
                stopSelf()
                return
            }
            try {
                if (needCookies)
                    Thread.sleep(6000)
                val document = jsoupCookiesDir("https://animeflv.net/browse?order=added&page=$page", needCookies).get()
                Log.e(TAG, "Read page $page")
                if (document.select("ul.ListAnimes").isNotEmpty() && document.select("article").isNotEmpty()) {
                    page++
                    val animeObjects = jspoon.adapter(DirectoryPage::class.java).fromHtml(document.outerHtml()).getAnimes(animeDAO, jspoon, object : DirectoryPage.UpdateInterface {
                        override fun onAdd() {
                            count++
                            updateNotification()
                        }

                        override fun onError() {
                            Log.e(TAG, "At page: $page")
                            if (strings?.contains(page.toString()) == false)
                                strings.add(page.toString())
                        }
                    }, needCookies)
                    if (animeObjects.isNotEmpty()) {
                        animeDAO.insertAll(animeObjects)
                    } else if (PrefsUtil.isDirectoryFinished || animeDAO.count >= maxAnimes) {
                        PrefsUtil.isDirectoryFinished = animeDAO.count >= maxAnimes
                        Log.e(TAG, "Stop searching at page $page")
                        cancelForeground()
                        break
                    }
                } else {
                    finished = true
                    Log.e(TAG, "Processed $page pages")
                    PrefsUtil.isDirectoryFinished = animeDAO.count >= maxAnimes
                    PreferenceManager.getDefaultSharedPreferences(this).edit().putStringSet(keyFailedPages, strings).apply()
                    DirUpdateWork.schedule(this)
                    setStatus(STATE_FINISHED)
                }
            } catch (e: HttpStatusException) {
                Log.e(TAG, "Page error: $page | Code: ${e.statusCode}")
                if (e.statusCode == 403 || e.statusCode == 503) {
                    e.printStackTrace()
                    finished = true
                    setStatus(STATE_INTERRUPTED)
                    stopSelf()
                    cancelForeground()
                }
            } catch (e: Exception) {
                Log.e(TAG, "Page error: $page | ${e.message}")
                if (strings?.contains(page.toString()) == false)
                    strings.add(page.toString())
                page++
                if (page > maxAnimes / 24){
                    finished = true
                    setStatus(STATE_INTERRUPTED)
                }
            }
        }
    }

    private fun cancelForeground() {
        noCrash {
            isRunning = false
            stopForeground(true)
            notCancel(NOT_CODE)
        }
    }

    private fun updateNotification() {
        val notification = NotificationCompat.Builder(this, CHANNEL).apply {
            setOngoing(true)
            priority = NotificationCompat.PRIORITY_MIN
            setSmallIcon(R.drawable.ic_directory_not)
            color = ContextCompat.getColor(this@DirectoryService, EAHelper.getThemeColor())
            setWhen(CURRENT_TIME)
            setSound(null)
            if (PrefsUtil.collapseDirectoryNotification)
                setSubText("Creando directorio: $count/$maxAnimes~")
            else {
                setContentTitle("Creando directorio")
                setContentText("Agregados: $count/$maxAnimes~")
                if (maxAnimes > 0)
                    setProgress(maxAnimes, count, false)
            }
        }
        notShow(NOT_CODE, notification.build())
    }

    private fun setStatus(status: Int) {
        doOnUI { liveStatus.setValue(status) }
    }

    private fun notShow(code: Int, notification: Notification) {
        manager?.notify(code, notification)
    }

    private fun notCancel(code: Int) {
        manager?.cancel(code)
    }

    override fun onTaskRemoved(rootIntent: Intent) {
        cancelForeground()
        super.onTaskRemoved(rootIntent)
    }

    interface OnDirStatus {
        fun onFinished()
    }

    companion object {
        const val STATE_PARTIAL = 0
        const val STATE_FULL = 1
        const val STATE_FINISHED = 2
        const val STATE_INTERRUPTED = 3
        const val STATE_VERIFYING = 4
        var NOT_CODE = 5598
        var CHANNEL = "directory_update"
        var isRunning = false
            private set
        private val liveStatus = MutableLiveData<Int>()

        fun run(context: Context?) {
            if (context == null) return
            if (!isRunning)
                context.service(Intent(context, DirectoryService::class.java))
        }

        fun getLiveStatus(): LiveData<Int> {
            return liveStatus
        }
    }
}
